### DISCUSSION TITLE
Issue with ssl protocol to connect to AWS (Solved)

### DISCUSSION THREAD

**Discussion by ebolisa** [micropython/micropython#14085] [RP2040 / Pico] [answered]
Hi,
I'd appreciate some directions on this issue. I'm trying to connect to an AWS server with certificates but, I get the error shown below. What am I doing wrong?
TIA

Picow:
```
MicroPython v1.22.2 on 2024-02-22; Raspberry Pi Pico W with RP2040
Libs used: simple and ssl
```

Error:
`Mqtt connection Error: unexpected keyword argument 'ssl_params'`

Code:
```
SSL_PARAMS = {
    "key": key,
    "cert": cert,
    "cadata": ca,
    "server_hostname": MQTT_BROKER,
    "cert_reqs": ssl.CERT_REQUIRED
    }


try:
    # create MQTT client that use TLS/SSL for a secure connection
    mqtt_client = MQTTClient(client_id=MQTT_CLIENT_ID,
                             server=MQTT_BROKER,
                             keepalive=60,
                             ssl=True,
                             ssl_params = SSL_PARAMS)
    
except Exception as e:
    print(f'Mqtt connection Error: {e}')
```

**Comment by aiot-maker** [comment, marked as answer]
I tested your code and it worked perfectly. So, probably there is something wrong on my side. Indeed the mqtt module simple.py that is installed in my Pico seems to be an old version. Apparently the MQTTClient Class definition has changed since I installed the module. in my test board.

class MQTTClient:

    def __init__(self, client_id, server, port=0, user=None, password=None, keepalive=0,
                 ssl=False, ssl_params={}):


versus latest Github

class MQTTClient:
    def __init__(
        self,
        client_id,
        server,
        port=0,
        user=None,
        password=None,
        keepalive=0,
        ssl=None,
    ):

**Comment by ebolisa** [comment]
Thank you @aiot-maker!!
I'm moving on but I'm getting this now:
```
Traceback (most recent call last):
  File "<stdin>", line 123, in <module>
  File "/lib/simple.py", line 69, in connect
AttributeError: 'bool' object has no attribute 'wrap_socket'
```
It seems my simple lib is out of phase. Where did you find yours?
Mine is from [here](https://github.com/micropython/micropython-lib/blob/master/micropython/umqtt.simple/umqtt/simple.py)
I also tried [this one](https://github.com/fizista/micropython-umqtt.simple2/blob/master/src/umqtt/simple2.py) where I get:
```
Traceback (most recent call last):
  File "<stdin>", line 123, in <module>
  File "/lib/simple2.py", line 267, in connect
OSError: [Errno 107] ENOTCONN
```
EDIT:
Solved!
Downloaded the lib from [here](https://pypi.org/project/micropython-umqtt.simple/#files)

**Comment by glenn20** [comment]
You can choose to use an older version of umqtt.simple OR you can use the recent versions which support SSLContext (see the [docs](https://docs.micropython.org/en/latest/library/ssl.html)).

You create an `SSLContext` instance, initialise it with the methods in the docs, and pass it as the `ssl` parameter of the MQTTClient constructor.

I recommend choosing the second approach ;).

**Reply by ebolisa** [reply]
Thanks but, where's the new lib located at?

**Reply by glenn20** [reply]
The latest is easily installed with `mpremote mip install umqtt.simple` if you have mpremote on your computer or from your device if it is internet connected `import mip; mip.install('umqtt.simple')`. See the [`mip` docs](https://docs.micropython.org/en/latest/reference/packages.html).

This will install from: https://github.com/micropython/micropython-lib/tree/master/micropython/umqtt.simple.

The micropython packages in pypi.org are generally no longer supported.

**Reply by ebolisa** [reply]
Thanks again! for some reason, **mpremote** doesn't work for me using Thonny. I'm following this [tut](https://www.hackster.io/sandeep-mistry/connect-your-raspberry-pi-pico-w-to-aws-iot-core-8868b7) which is using this [code](https://www.hackster.io/sandeep-mistry/connect-your-raspberry-pi-pico-w-to-aws-iot-core-8868b7) and both lines point to the same **url**.
So, if we look at the url's page we'll find the **simple.py** lib. However, that lib (which is what I used initially) doesn't have **ssl_params={}** in their init line **def __init__(self,client_id,server,port=0,user=None,password=None,keepalive=0,ssl=None,)**.
Surely, I'm missing something here.

**Comment by AndresG25** [comment]
Hello! I'd like to know if someone could solve the problem. I'm actually working with a nanoESP32-C6 and I'm trying to send data from the ESP to AWS IoT Core, I'm programming with MicroPython and my ESP doesn´t have support with CircuitPython yet. What can I do? Thanks for your help! Excuse me for my English, I'm from Colombia. My error: unexpected keyword argument 'ssl_params'

**Comment by ebolisa** [comment]
How are you formatting the parameters' code?

Here's an example:

```
SSL_PARAMS = {
    "key": key,
    "cert": cert,
    "cadata": ca,
    "server_hostname": MQTT_BROKER,
    "cert_reqs": ssl.CERT_REQUIRED
    }

try:
    # create MQTT client that use TLS/SSL for a secure connection
    mqtt_client = MQTTClient(client_id=MQTT_CLIENT_ID,
                             server=MQTT_BROKER,
                             keepalive=60,
                             ssl=True,
                             ssl_params = SSL_PARAMS)
```

**Reply by AndresG25** [reply]
I got to solve it with the library that you published: https://pypi.org/project/micropython-umqtt.simple/#files 
However, I'm getting a new error with my ESP32C6 with Micropython: no module named 'ussl'

Do you know why is this problem?

**Reply by ebolisa** [reply]
> I got to solve it with the library that you published: https://pypi.org/project/micropython-umqtt.simple/#files However, I'm getting a new error with my ESP32C6 with Micropython: no module named 'ussl'
> 
> Do you know why is this problem?

Actually, this' an old thread. I'm using now [this](https://github.com/peterhinch/micropython-mqtt) lib. 
As for your error, type help ('modules') to see if you have it. Otherwise, you can always do this:
```
try:
    import ussl
except:
   import ssl
```
